lib/Sema/DerivedConformanceTensorGroup.cpp

```diff
@@ -0,0 +1,396 @@
+  auto plusOpDRE = new (C) 
+      DeclRefExpr(plusOpDecl, DeclNameLoc(), /*Implicit*/ true);
+  auto plusOpExpr = 
+        new (C) DotSyntaxCallExpr(plusOpDRE, SourceLoc(), arrayTypeExpr);
+  Expr *typeListExpr = nullptr;
+  for (auto member : nominal->getStoredProperties()) {
+    auto memberType =
+        parentDC->mapTypeIntoContext(member->getValueInterfaceType());
+    auto *memberTypeExpr = TypeExpr::createImplicit(memberType, C);
+    auto *memberTypeListExpr = new (C) 
+        MemberRefExpr(memberTypeExpr, SourceLoc(), typeListReq,
+                      DeclNameLoc(), /*Implicit*/ true);
+    if (typeListExpr == nullptr)  {
+      typeListExpr = memberTypeListExpr;
+    } else {
+      // Create expression `lhsArg + rhsArg`.
+    auto *plusOpArgs =
+        TupleExpr::create(C, SourceLoc(), {typeListExpr, memberTypeListExpr}, 
+                          {}, {}, SourceLoc(), /*HasTrailingClosure*/ false,
+                          /*Implicit*/ true);
+    typeListExpr = new (C) BinaryExpr(plusOpExpr, plusOpArgs, 
+                                      /*Implicit*/ true);
+    }
+  }
+
+  // Return the resulting data types array.
+  auto *returnStmt = new (C) ReturnStmt(SourceLoc(), typeListExpr);
+  auto *body = BraceStmt::create(C, SourceLoc(), {returnStmt}, SourceLoc(),
+                                 /*Implicit*/ true);
+  funcDecl->setBody(BraceStmt::create(C, SourceLoc(), {body}, SourceLoc(),
+                                      /*Implicit*/ true));
+}
+
+/// Derive a '_typeList' implementation.
+static ValueDecl *deriveTensorGroup_typeList(DerivedConformance &derived) {
+  auto nominal = derived.Nominal;
+  auto &TC = derived.TC;
+  ASTContext &C = TC.Context;
+
+  auto parentDC = derived.getConformanceContext();
+  Type dataTypeArrayType = BoundGenericType::get(
+    C.getArrayDecl(), Type(), 
+    {C.getTensorDataTypeDecl()->getDeclaredInterfaceType()});
+  auto returnType = parentDC->mapTypeIntoContext(dataTypeArrayType);
+
+  // Create `_typeList` property declaration.
+  VarDecl *typeListDecl;
+  PatternBindingDecl *patDecl;
+  std::tie(typeListDecl, patDecl) = derived.declareDerivedProperty(
+      C.Id_typeList, returnType, returnType, /*isStatic*/ true,
+      /*isFinal*/ false);
+
+  // Add `@inlinable` to the `_typeList` declaration.
+  if (nominal->getEffectiveAccess() > AccessLevel::Internal)
+    typeListDecl->getAttrs().add(new (C) InlinableAttr(/*implicit*/ true));
+
+  // Create `_typeList` getter.
+  auto *getterDecl = derived.declareDerivedPropertyGetter(
+      TC, typeListDecl, returnType);
+  getterDecl->setBodySynthesizer(deriveBodyTensorGroup_typeList);
+  typeListDecl->setAccessors(StorageImplInfo::getImmutableComputed(),
+                                SourceLoc(), {getterDecl}, SourceLoc());
+  derived.addMembersToConformanceContext({getterDecl, typeListDecl, patDecl});
+
+  return typeListDecl;
+}
+
+// Synthesize body for `init(_owning:)`.
+static void 
+deriveBodyTensorGroup_init(
```

Comment:
Combine into one line:
```suggestion
deriveBodyTensorGroup_init(AbstractFunctionDecl *funcDecl) {
```